src/com/ms/silverking/cloud/topology/TopologyParser.java

```diff
@@ -11,157 +11,158 @@
-                                case ENTRY:
-                                    root = new TopologyEntry(token.getEntry());
-                                    parentEntries.add(root);
-                                    break;
-                                default: throw new RuntimeException("panic");
-                                }
-                            } else {
-                                switch (token.getType()) {
-                                case OPEN_BLOCK:
-                                    if (prevToken != null) {
-                                        switch (prevToken.getType()) {
-                                        case OPEN_BLOCK:
-                                        case CLOSE_BLOCK:
-                                            throw new TopologyParseException("Invalid block defiition");
-                                        case ENTRY:
-                                            TopologyEntry    parent;
-                                            TopologyEntry    lastAdded;
-                                            
-                                            parent = parentEntries.get(parentEntries.size() - 1);
-                                            lastAdded = parent.getLastChild();
-                                            parentEntries.add(lastAdded);
-                                            break;
-                                        default: throw new RuntimeException("panic");
-                                        }
-                                    }
-                                    break;
-                                case CLOSE_BLOCK:
-                                    switch (prevToken.getType()) {
-                                    case CLOSE_BLOCK:
-                                    case OPEN_BLOCK:
-                                    case ENTRY:
-                                        parentEntries.remove(parentEntries.size() - 1);
-                                        break;
-                                    default: throw new RuntimeException("panic");
-                                    }
-                                    break;
-                                case ENTRY:
-                                    parentEntries.get(parentEntries.size() - 1).addChild(new TopologyEntry(token.getEntry()));
-                                    break;
-                                default: throw new RuntimeException("panic");
-                                }
-                                prevToken = token;
-                            }
-                        }
+  private static final long noVersion = Long.MAX_VALUE;
+
+  private static final boolean debug = false;
+
+  public TopologyParser() {
+  }
+
+  public static Topology parse(File file) throws IOException, TopologyParseException {
+    return parse(new FileInputStream(file));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource created by the following constructor call: *FileInputStream*. Currently, there are execution paths that do not contain closure statements, for example, when TopologyParser.parse throws an exception. Either a) close the object returned by *FileInputStream()* in a try-finally block or b) close the resource by declaring the object returned by *FileInputStream()* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).